openfermioncirq/gates/four_qubit_gates.py

```diff
@@ -0,0 +1,177 @@
+#   limitations under the License.
+
+"""Gates that target four qubits."""
+
+from typing import Optional, Union
+
+import numpy
+
+import cirq
+
+
+class LocalPQRSGate(cirq.EigenGate,
+                    cirq.CompositeGate,
+                    cirq.TextDiagrammable):
+    """Evolve under |0011><1100| + h.c. for some time."""
+
+    def __init__(self, *,  # Forces keyword args.
+                 half_turns: Optional[Union[cirq.Symbol, float]]=None,
+                 rads: Optional[float]=None,
+                 degs: Optional[float]=None,
+                 duration: Optional[float]=None) -> None:
+        """Initialize the gate.
+
+        At most one of half_turns, rads, degs, or duration may be specified.
+        If more are specified, the result is considered ambiguous and an
+        error is thrown. If no argument is given, the default value of one
+        half-turn is used.
+
+        Args:
+            half_turns: The exponent angle, in half-turns.
+            rads: The exponent angle, in radians.
+            degs: The exponent angle, in degrees.
+            duration: The exponent as a duration of time.
+        """
+
+        if len([1 for e in [half_turns, rads, degs, duration]
+                if e is not None]) > 1:
+            raise ValueError('Redundant exponent specification. '
+                             'Use ONE of half_turns, rads, degs, or duration.')
+
+        if duration is not None:
+            exponent = 2 * duration / numpy.pi
+        else:
+            exponent = cirq.value.chosen_angle_to_half_turns(
+                half_turns=half_turns,
+                rads=rads,
+                degs=degs)
+
+        super().__init__(exponent=exponent)
+
+    @property
+    def half_turns(self) -> Union[cirq.Symbol, float]:
+        return self._exponent
+
+    def _eigen_components(self):
+        return [
+            (-1, numpy.array(
+                [[0, 0, 0, 0, 0, 0, 0, 0,
+                  0, 0, 0, 0, 0, 0, 0, 0] for i in range(3)] +
+                [[0, 0, 0, 0.5, 0, 0, 0, 0,
+                  0, 0, 0, 0, 0.5, 0, 0, 0]] +
+                [[0, 0, 0, 0, 0, 0, 0, 0,
+                  0, 0, 0, 0, 0, 0, 0, 0] for i in range(8)] +
+                [[0, 0, 0, 0.5, 0, 0, 0, 0,
+                  0, 0, 0, 0, 0.5, 0, 0, 0]] +
+                [[0, 0, 0, 0, 0, 0, 0, 0,
+                  0, 0, 0, 0, 0, 0, 0, 0] for i in range(3)])),
+            (1, numpy.array(
+                [[0, 0, 0, 0, 0, 0, 0, 0,
+                  0, 0, 0, 0, 0, 0, 0, 0] for i in range(3)] +
+                [[0, 0, 0, 0.5, 0, 0, 0, 0,
+                  0, 0, 0, 0, -0.5, 0, 0, 0]] +
+                [[0, 0, 0, 0, 0, 0, 0, 0,
+                  0, 0, 0, 0, 0, 0, 0, 0] for i in range(8)] +
+                [[0, 0, 0, -0.5, 0, 0, 0, 0,
+                  0, 0, 0, 0, 0.5, 0, 0, 0]] +
+                [[0, 0, 0, 0, 0, 0, 0, 0,
+                  0, 0, 0, 0, 0, 0, 0, 0] for i in range(3)])),
```

Comment:
This is unreadable. Initialize a numpy array of zeros and directly set the appropriate indices to the appropriate values.